leader-election-impl/src/main/java/com/palantir/leader/PaxosLeaderElectionService.java

```diff
@@ -194,14 +260,13 @@ private boolean pingLeader() {
         }
         final PingableLeader leader = maybeLeader.get();
 
-        CompletionService<Boolean> pingCompletionService = new ExecutorCompletionService<Boolean>(
-                executor);
+        MultiplexingCompletionService<PingableLeader, Boolean> multiplexingCompletionService
```

Comment:
why do we always create a new one? don't we lose progress between calls then?

Reply:
+ aren't we then creating a lot of thread pools?

Reply:
This something I am also concerned with. We won't be losing progress, but not sure about the resource cost of creating this object again and again. I run a few load tests to see if it has an effect on perf, and it looks fine - but I would also want to hear @jeremyk-91 's thoughts here.

Reply:
Regarding thread pools, we aren't re-creating them; the same set of executors (and hence same underlying thread pools) are being used even in separate instances of the MCS.

Think of MCS as a bridge that sends a task to some executors, and then has a queue where it collects the responses - the executors themselves aren't a part of an MCS. There might be some overhead for creating MCS and its completion queue, but the underlying executors are the same (thanks for load testing @gokcanozakdag!)

We have to create a separate MCS per request because responses across requests aren't generally interchangeable (e.g. a `getSuspectedLeaderOverNetwork()` invoked with different string arguments). For ping(), it may be possible to re-use responses in some way - though I didn't intend that to be part of this PR.